operators/src/processing/expression.rs

```diff
@@ -0,0 +1,361 @@
+    for InitializedOperatorImpl<ExpressionParams, RasterResultDescriptor, ()>
+{
+    fn query_processor(&self) -> Result<TypedRasterQueryProcessor> {
+        // TODO: handle different number of sources
+
+        match *self.raster_sources.as_slice() {
+            [ref a, ref b] => {
+                let a = a.query_processor()?;
+                let b = b.query_processor()?;
+
+                call_bi_generic_processor!(a, b, (p_a, p_b) => {
+                    let res = call_generic_raster_processor!(self.params.output_type, ExpressionQueryProcessor::new(
+                                &SafeExpression {
+                                    expression: self.params.expression.clone(),
+                                },
+                                p_a,
+                                p_b,
+                            )
+                            .boxed());
+                    Ok(res)
+                })
+            }
+            _ => unimplemented!(), // TODO: handle more than two inputs
+        }
+    }
+}
+
+struct ExpressionQueryProcessor<T1, T2, TO>
+where
+    T1: Pixel,
+    T2: Pixel,
+    TO: Pixel,
+{
+    pub source_a: Box<dyn RasterQueryProcessor<RasterType = T1>>,
+    pub source_b: Box<dyn RasterQueryProcessor<RasterType = T2>>,
+    pub phantom_data: PhantomData<TO>,
+    pub cl_program: CompiledCLProgram,
+}
+
+impl<T1, T2, TO> ExpressionQueryProcessor<T1, T2, TO>
+where
+    T1: Pixel,
+    T2: Pixel,
+    TO: Pixel,
+{
+    fn new(
+        expression: &SafeExpression,
+        source_a: Box<dyn RasterQueryProcessor<RasterType = T1>>,
+        source_b: Box<dyn RasterQueryProcessor<RasterType = T2>>,
+    ) -> Self {
+        Self {
+            source_a,
+            source_b,
+            cl_program: Self::create_cl_program(&expression),
+            phantom_data: PhantomData::default(),
+        }
+    }
+
+    fn create_cl_program(expression: &SafeExpression) -> CompiledCLProgram {
+        // TODO: generate code for arbitrary amount of inputs
+        let source = r#"
+__kernel void expressionkernel(
+            __global const IN_TYPE0 *in_data0,
+            __global const RasterInfo *in_info0,
+            __global const IN_TYPE1* in_data1,
+            __global const RasterInfo *in_info1,
+            __global OUT_TYPE0* out_data,
+            __global const RasterInfo *out_info)
+{
+    uint const gid = get_global_id(0) + get_global_id(1) * in_info0->size[0];
+    if (gid >= in_info0->size[0]*in_info0->size[1]*in_info0->size[2])
+        return;
+
+    IN_TYPE0 A = in_data0[gid];
+    if (ISNODATA0(A, in_info0)) {
+        out_data[gid] = out_info->no_data;
+        return;
+    }
+
+    IN_TYPE0 B = in_data1[gid];
+    if (ISNODATA1(B, in_info1)) {
+        out_data[gid] = out_info->no_data;
+        return;
+    }
+
+    OUT_TYPE0 result = %%%EXPRESSION%%%;
+	out_data[gid] = result;
+}"#
+        .replace("%%%EXPRESSION%%%", &expression.expression);
```

Comment:
Why don't you use a format string?

Reply:
because that would require escaping all the curly braces and make the code worse the read